src/assignment-4/FractionalWrapper.sol

```diff
@@ -0,0 +1,75 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.13;
+
+import {ERC20Permit} from "yield-utils-v2/token/ERC20Permit.sol";
+import {IERC20} from "yield-utils-v2/token/IERC20.sol";
+import {IYearnVault} from "./YearnVaultMock.sol";
+
+/// @title FractionalWrapper is a thin ERC20 wrapper over a Yearn Vault
+/// @author davidbrai
+/// @notice The wrapper accepts deposits of the underlying token, and in turn,
+///     deposits them into the Yearn Vault.
+///     The user gets back FractionalWrapper shares which are 1:1 representation of the Yearn Vault shares.
+///     The user can burn the shares back into the wrapper, which will burn the Yearn Vault shares, and return
+///     the underlying tokens returned from the Yearn Vault.
+contract FractionalWrapper is ERC20Permit {
+
+    IERC20 public immutable token;
+    IYearnVault public immutable yvToken;
+
+    error TransferFailed();
+
+    /// @notice Initializes a new FractionalWrapper
+    /// @param token_ The underlying token
+    /// @param yvToken_ The Yearn Vault being wrapped
+    /// @param name The name of the wrapper
+    /// @param symbol The symbol of the wrapper
+    /// @param decimals The number of decimals for the wrapper, should be the same as the Yearn Vault
+    constructor(address token_, address yvToken_, string memory name, string memory symbol, uint8 decimals) ERC20Permit(name, symbol, decimals) {
+        token = IERC20(token_);
+        yvToken = IYearnVault(yvToken_);
+
+        // Allow Yearn Vault to transfer underlying token from the wrapper
+        token.approve(yvToken_, type(uint256).max);
+    }
+
+    /// @notice Deposit tokens into the contract and gets back wrapper shares which wrap Yearn Vault shares
+    ///     1. User sends `amount` of `token` to wrapper
+    ///     2. Wrapper sends `amount` of `token` to Yearn Vault
+    ///     3. Yearn Vault sends `numShares` of `yvToken` to wrapper
+    ///     4. Wrapper sends `numShares` of wrapper shares to user
+    /// @param amount The amount of underlying token to deposit
+    function deposit(uint amount) public {
```

Comment:
It's good practice to always return any values that are obtained during execution, in this case `numShares`. It helps enormously to integrate this contract with others.

Reply:
makes sense. fixed here: https://github.com/davidbrai/yield-mentorship/pull/5/commits/f67495b08eca3daa7a8f576a1386d694cf707e3c